### DISCUSSION TITLE
Rust初探: 实现一个Ping

### DISCUSSION THREAD

**Discussion by giscus** [qingwave/qingwave.github.io#89] [Announcements]
# Rust初探: 实现一个Ping

Rust + Ping = Ring

https://qingwave.github.io/rust-ping/

**Comment by stallonefennec** [comment]
111

**Comment by feynhang** [comment]
Socket::new(Domain::IPV4, Type::DGRAM, Some(Protocol::ICMPV4)) 这里必定返回错误 “请求的协议还没有在系统中配置，或者没有它存在的迹象。 (os error 10043)”，这里Type用错了吧？ICMP是网络层协议不能使用udp，udp属于传输层比icmp要高一层（依赖端口号），明显冲突了。你这怎么跑通的？

**Reply by qingwave** [reply]
感谢指正，应该用type:raw，但确实可以正常运行，应该是底层操作系统会做一些默认转换吧

**Reply by feynhang** [reply]
我最近又去看了几个实现，发现使用`Type::DGRAM`初始化socket目的是要实现unprivileged icmp， 然而这种实现似乎仅限类unix平台上，Windows对unprivilege icmp的实现必须使用的单独win32 API（Icmpapi.h），使用`Type::DGRAM`初始化的`socket`这种通用接口似乎行不通，反正我这台式机PC上在`socket::new`这直接返回错误了，后面都不会运行。

**Reply by qingwave** [reply]
想起来了，当时实现的时候确实因为`Type::RAW`需要特权，才使用的`Type::DGRAM`，可以参考[add ICMP socket kind](https://lwn.net/Articles/420800/)，刚也发现一个更完整的关于`DGRAM`[兼容性的测试](https://github.com/kolapapa/surge-ping/issues/30#issuecomment-1314704664)，目前Linux和MacOS是支持的。
